base/XGamepad.ts

```diff
@@ -0,0 +1,503 @@
+                    const buttons = Array.prototype.filter.call(matches, (m, i) => i && m);
+                    km.button_codes = buttons.map(get_button_code).filter(bc => bc !== undefined);
+                    km.game_buttons = buttons.map(get_game_button).filter(bc => bc !== undefined);
+                    if (km.button_codes.length !== km.game_buttons.length)
+                        console.warn(`${button_code} not well recognized!`);
+                } else km.button_code = get_button_code(matches[matches.length - 1]);
+                return km;
+            }
+        );
+        // Load gamepad stick configuration
+        const gamepad_stick_mapping = Object.entries(
+            data.dbs.init_db.gamepad_rsticks as {[code: string]: [string, string]}
+        ).map(
+            ([button_code, game_buttons]): KeyMap => ({
+                game_button: null,
+                game_buttons: game_buttons.map(gb => CButton[gb]),
+                button_code: Phaser.Gamepad[button_code],
+            })
+        );
+        Gamepad.gamepad_stick_mapping = gamepad_stick_mapping;
+        Gamepad.gamepad_mapping = gamepad_mapping.concat(gamepad_custom_mapping);
+        // Load keyboard keys configuration
+        Gamepad.keyboard_mapping = Object.entries(data.dbs.init_db[input_type] as {[code: string]: string}).map(
+            ([game_button, key_code]): KeyMap => {
+                const matches = key_code.match(
+                    /^(?:(?:(?:(ALT|CTRL|SHIFT) *\+ *)?(ALT|CTRL|SHIFT) *\+ *)?(ALT|CTRL|SHIFT) *\+ *)?(\w+)$/
+                );
+                if (!matches) console.error("Input not recognized " + key_code);
+                const km: KeyMap = {
+                    name: game_button,
+                    as: key_code,
+                    game_button: Button[game_button] ?? CButton[game_button],
+                    key_code: Phaser.Keyboard[matches[matches.length - 1]],
+                };
+                if (matches[matches.length - 2]) {
+                    const modifiers = Array.prototype.filter.call(matches, (m, i) => i && m);
+                    km.key_modifiers = {};
+                    if (modifiers.includes("ALT")) km.key_modifiers.alt = true;
+                    if (modifiers.includes("CTRL")) km.key_modifiers.ctrl = true;
+                    if (modifiers.includes("SHIFT")) km.key_modifiers.shift = true;
+                }
+                return km;
+            }
+        );
+    }
+
+    /**
+     * Gets the GBA button(s) attached to the controller button.
+     * @param {number} button_code - Controller button code.
+     * @return {(Button|CButton[]} - GBA (custom) button(s)
+     */
+    static transcode_gamepad_button(button_code: number) {
+        // return Gamepad.keyboard_mapping.find(km => km.button_code === button_code)?.game_button;
+        return Gamepad.gamepad_mapping.filter(km => km.button_code === button_code).map(km => km.game_button);
```

Comment:
Not a problem at all, but consider using `flatMap` instead of  `map` + `filter` next time.

Reply:
`.flatMap` is much slower than `.map` + `.filter` and i find it less readable.